File: composeApp/src/main/java/com/wotty/stark/data/local/LocalRepository.kt
package com.wotty.stark.data.local

import android.content.Context
import com.wotty.stark.data.model.*
import com.wotty.stark.data.repository.DataRepository
import com.wotty.stark.data.repository.ImportResult
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import java.io.File

/**
 * 模式1：本地 JSON 文件存储
 */
class LocalRepository(context: Context) : DataRepository {

    private val json = Json {
        ignoreUnknownKeys = true
        prettyPrint = true
        coerceInputValues = true
    }

    private val dataDir = File(context.filesDir, "data").also { it.mkdirs() }

    private inline fun <reified T> readJson(filename: String, serializer: kotlinx.serialization.KSerializer<T>): T? {
        val file = File(dataDir, filename)
        if (!file.exists()) return null
        return json.decodeFromString(serializer, file.readText())
    }

    private inline fun <reified T> writeJson(filename: String, data: T, serializer: kotlinx.serialization.KSerializer<T>) {
        File(dataDir, filename).writeText(json.encodeToString(serializer, data))
    }

    private inline fun <reified T> readListJson(filename: String, serializer: kotlinx.serialization.KSerializer<T>): List<T> {
        val file = File(dataDir, filename)
        if (!file.exists()) return emptyList()
        return json.decodeFromString(ListSerializer(serializer), file.readText())
    }

    private inline fun <reified T> writeListJson(filename: String, data: List<T>, serializer: kotlinx.serialization.KSerializer<T>) {
        File(dataDir, filename).writeText(json.encodeToString(ListSerializer(serializer), data))
    }

    override suspend fun getCurrentUser(): User? = readJson("user.json", User.serializer())
    override suspend fun saveUser(user: User) = writeJson("user.json", user, User.serializer())

    override suspend fun getAccounts(): List<Account> = readListJson("account.json", Account.serializer())
    override suspend fun getAccount(id: String): Account? = getAccounts().find { it.id == id }
    override suspend fun saveAccount(account: Account) {
        val list = getAccounts().toMutableList()
        val idx = list.indexOfFirst { it.id == account.id }
        if (idx >= 0) list[idx] = account else list.add(account)
        writeListJson("account.json", list, Account.serializer())
    }
    override suspend fun deleteAccount(id: String) {
        writeListJson("account.json", getAccounts().filter { it.id != id }, Account.serializer())
    }

    override suspend fun getTransactions(accountId: String, page: Int, pageSize: Int): List<Transaction> {
        val all = readListJson("transaction.json", Transaction.serializer())
            .filter { it.accountId == accountId }
            .sortedByDescending { it.date }
        return all.drop((page - 1) * pageSize).take(pageSize)
    }
    override suspend fun getTransaction(id: String): Transaction? =
        readListJson("transaction.json", Transaction.serializer()).find { it.id == id }
    override suspend fun saveTransaction(transaction: Transaction) {
        val list = readListJson("transaction.json", Transaction.serializer()).toMutableList()
        val idx = list.indexOfFirst { it.id == transaction.id }
        if (idx >= 0) list[idx] = transaction else list.add(transaction)
        writeListJson("transaction.json", list, Transaction.serializer())
    }
    override suspend fun deleteTransaction(id: String) {
        writeListJson("transaction.json",
            readListJson("transaction.json", Transaction.serializer()).filter { it.id != id },
            Transaction.serializer())
    }
    override suspend fun importTransactions(transactions: List<Transaction>): ImportResult {
        val existing = readListJson("transaction.json", Transaction.serializer()).toMutableList()
        var imported = 0; var skipped = 0
        transactions.forEach { t ->
            val dup = t.orderId != null && existing.any { it.orderId == t.orderId }
            if (dup) skipped++ else { existing.add(t); imported++ }
        }
        writeListJson("transaction.json", existing, Transaction.serializer())
        return ImportResult(imported = imported, skipped = skipped)
    }

    override suspend fun getAssets(accountId: String): List<Asset> =
        readListJson("asset.json", Asset.serializer()).filter { it.accountId == accountId }
    override suspend fun saveAsset(asset: Asset) {
        val list = readListJson("asset.json", Asset.serializer()).toMutableList()
        val idx = list.indexOfFirst { it.id == asset.id }
        if (idx >= 0) list[idx] = asset else list.add(asset)
        writeListJson("asset.json", list, Asset.serializer())
    }
    override suspend fun deleteAsset(id: String) {
        writeListJson("asset.json", readListJson("asset.json", Asset.serializer()).filter { it.id != id }, Asset.serializer())
    }

    override suspend fun getBudgets(accountId: String): List<Budget> =
        readListJson("budget.json", Budget.serializer()).filter { it.accountId == accountId }
    override suspend fun saveBudget(budget: Budget) {
        val list = readListJson("budget.json", Budget.serializer()).toMutableList()
        val idx = list.indexOfFirst { it.id == budget.id }
        if (idx >= 0) list[idx] = budget else list.add(budget)
        writeListJson("budget.json", list, Budget.serializer())
    }
    override suspend fun deleteBudget(id: String) {
        writeListJson("budget.json", readListJson("budget.json", Budget.serializer()).filter { it.id != id }, Budget.serializer())
    }

    override suspend fun getLoans(accountId: String): List<Loan> =
        readListJson("loan.json", Loan.serializer()).filter { it.accountId == accountId }
    override suspend fun saveLoan(loan: Loan) {
        val list = readListJson("loan.json", Loan.serializer()).toMutableList()
        val idx = list.indexOfFirst { it.id == loan.id }
        if (idx >= 0) list[idx] = loan else list.add(loan)
        writeListJson("loan.json", list, Loan.serializer())
    }
    override suspend fun deleteLoan(id: String) {
        writeListJson("loan.json", readListJson("loan.json", Loan.serializer()).filter { it.id != id }, Loan.serializer())
    }

    override suspend fun getSavingsGoals(accountId: String): List<SavingsGoal> =
        readListJson("savingsgoal.json", SavingsGoal.serializer()).filter { it.accountId == accountId }
    override suspend fun saveSavingsGoal(goal: SavingsGoal) {
        val list = readListJson("savingsgoal.json", SavingsGoal.serializer()).toMutableList()
        val idx = list.indexOfFirst { it.id == goal.id }
        if (idx >= 0) list[idx] = goal else list.add(goal)
        writeListJson("savingsgoal.json", list, SavingsGoal.serializer())
    }
    override suspend fun deleteSavingsGoal(id: String) {
        writeListJson("savingsgoal.json", readListJson("savingsgoal.json", SavingsGoal.serializer()).filter { it.id != id }, SavingsGoal.serializer())
    }
    override suspend fun getSavingsPlans(goalId: String): List<SavingsPlan> =
        readListJson("savingsplan.json", SavingsPlan.serializer()).filter { it.goalId == goalId }
    override suspend fun saveSavingsPlan(plan: SavingsPlan) {
        val list = readListJson("savingsplan.json", SavingsPlan.serializer()).toMutableList()
        val idx = list.indexOfFirst { it.id == plan.id }
        if (idx >= 0) list[idx] = plan else list.add(plan)
        writeListJson("savingsplan.json", list, SavingsPlan.serializer())
    }

    override suspend fun getCategoryRules(accountId: String): List<CategoryRule> =
        readListJson("category_rules.json", CategoryRule.serializer()).filter { it.accountId == accountId }
    override suspend fun saveCategoryRule(rule: CategoryRule) {
        val list = readListJson("category_rules.json", CategoryRule.serializer()).toMutableList()
        val idx = list.indexOfFirst { it.id == rule.id }
        if (idx >= 0) list[idx] = rule else list.add(rule)
        writeListJson("category_rules.json", list, CategoryRule.serializer())
    }

    override suspend fun getImportErrorLogs(accountId: String): List<ImportErrorLog> =
        readListJson("importerrorlog.json", ImportErrorLog.serializer()).filter { it.accountId == accountId }
    override suspend fun saveImportErrorLog(log: ImportErrorLog) {
        val list = readListJson("importerrorlog.json", ImportErrorLog.serializer()).toMutableList()
        list.add(log)
        writeListJson("importerrorlog.json", list, ImportErrorLog.serializer())
    }

    override suspend fun getExchangeRates(): List<ExchangeRate> =
        readListJson("exchangerate.json", ExchangeRate.serializer())

    override suspend fun getThemeConfig(userId: String): ThemeConfig? =
        readListJson("themeconfig.json", ThemeConfig.serializer()).firstOrNull { it.userId == userId }
    override suspend fun saveThemeConfig(config: ThemeConfig) {
        val list = readListJson("themeconfig.json", ThemeConfig.serializer()).toMutableList()
        val idx = list.indexOfFirst { it.id == config.id }
        if (idx >= 0) list[idx] = config else list.add(config)
        writeListJson("themeconfig.json", list, ThemeConfig.serializer())
    }

    override suspend fun seedDemoData() {
        val now = java.time.LocalDateTime.now().format(java.time.format.DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss"))
        val month = now.take(7)
        val txns = listOf(
            Transaction("txn-demo-salary", "local", "default", 12800.0, TransactionType.INCOME, "工资", "银行卡", "公司发薪", "${month}-15 09:00:00", "月中发薪", createdAt = now, updatedAt = now),
            Transaction("txn-demo-food", "local", "default", 86.0, TransactionType.EXPENSE, "餐饮", "支付宝", "午餐", "${month}-16 12:15:00", "工作餐", createdAt = now, updatedAt = now),
            Transaction("txn-demo-shop", "local", "default", 268.0, TransactionType.EXPENSE, "购物", "微信", "日用品", "${month}-18 20:10:00", "家庭采购", createdAt = now, updatedAt = now),
            Transaction("txn-demo-traffic", "local", "default", 48.0, TransactionType.EXPENSE, "交通", "支付宝", "地铁公交", "${month}-19 08:30:00", "通勤", createdAt = now, updatedAt = now),
            Transaction("txn-demo-entertain", "local", "default", 156.0, TransactionType.EXPENSE, "娱乐", "微信", "电影票", "${month}-20 20:40:00", "周末观影", createdAt = now, updatedAt = now)
        )
        val assets = listOf(
            Asset("asset-demo-bank", "local", "default", "工资卡", AssetType.BANK_CARD, 48216.4, createdAt = now, updatedAt = now),
            Asset("asset-demo-wechat", "local", "default", "微信钱包", AssetType.WECHAT, 1260.5, createdAt = now, updatedAt = now),
            Asset("asset-demo-alipay", "local", "default", "支付宝余额", AssetType.ALIPAY, 3180.0, createdAt = now, updatedAt = now)
        )
        val budgets = listOf(
            Budget("budget-demo-global", "local", "default", 6000.0, "ALL", alertPercent = 80, scopeType = BudgetScopeType.GLOBAL, createdAt = now, updatedAt = now),
            Budget("budget-demo-food", "local", "default", 1500.0, "餐饮", alertPercent = 80, scopeType = BudgetScopeType.CATEGORY, createdAt = now, updatedAt = now)
        )
        val loans = listOf(
            Loan("loan-demo-home", "local", "default", "房贷", 480000.0, 352000.0, 240, 64, 3200.0, 20, createdAt = now, updatedAt = now),
            Loan("loan-demo-car", "local", "default", "车贷", 80000.0, 27000.0, 36, 18, 2200.0, 10, createdAt = now, updatedAt = now)
        )
        val goals = listOf(
            SavingsGoal("goal-demo-travel", "local", "default", "旅行基金", 30000.0, 9200.0, "2026-12-31", SavingsGoalType.LONG_TERM, SavingsGoalStatus.ACTIVE, SavingsGoalDepositType.CASH, createdAt = now, updatedAt = now),
            SavingsGoal("goal-demo-emergency", "local", "default", "应急储备", 20000.0, 6800.0, "2027-06-30", SavingsGoalType.LONG_TERM, SavingsGoalStatus.ACTIVE, SavingsGoalDepositType.CASH, createdAt = now, updatedAt = now)
        )
        val plans = listOf(
            SavingsPlan("plan-demo-travel-1", "goal-demo-travel", 2000.0, SavingsPlanStatus.PENDING, month, createdAt = now, updatedAt = now)
        )
        txns.forEach { saveTransaction(it) }
        assets.forEach { saveAsset(it) }
        budgets.forEach { saveBudget(it) }
        loans.forEach { saveLoan(it) }
        goals.forEach { saveSavingsGoal(it) }
        plans.forEach { saveSavingsPlan(it) }
    }
}

File: web/src/lib/stark/repository/LocalRepository.ts
import type {
  Account,
  Asset,
  Budget,
  CategoryRule,
  ExchangeRate,
  ImportErrorLog,
  ImportResult,
  Loan,
  SavingsGoal,
  SavingsPlan,
  ThemeConfig,
  Transaction,
  User,
} from "@/lib/stark/models/types";
import type { DataRepository } from "@/lib/stark/repository/DataRepository";
import { deleteRecord, getAllRecords, getRecord, putManyRecords, putRecord } from "@/lib/stark/storage/indexeddb";
import { getCurrentAccountId } from "@/lib/stark/storage/local-config";
import { nowText } from "@/lib/stark/utils/format";

function uuid() {
  if (typeof crypto !== "undefined" && typeof crypto.randomUUID === "function") {
    return crypto.randomUUID();
  }
  return `${Date.now()}-${Math.random().toString(16).slice(2)}`;
}

function sortByDateDesc<T extends { date?: string; createdAt?: string }>(items: T[]) {
  return [...items].sort((a, b) => {
    const aValue = a.date ?? a.createdAt ?? "";
    const bValue = b.date ?? b.createdAt ?? "";
    return bValue.localeCompare(aValue);
  });
}

function defaultUser(): User {
  const now = nowText();
  return {
    id: "local-user",
    email: "local@wotty.stark",
    password: "",
    name: "本地用户",
    defaultAccountId: "default",
    role: "USER",
    createdAt: now,
    updatedAt: now,
  };
}

function defaultAccount(): Account {
  const now = nowText();
  return {
    id: "default",
    name: "默认账本",
    ownerId: "local-user",
    createdAt: now,
    updatedAt: now,
  };
}

export class LocalRepository implements DataRepository {
  private seeded: Promise<void> | null = null;

  private async ensureSeeded() {
    if (!this.seeded) {
      this.seeded = this.seed();
    }
    return this.seeded;
  }

  private async seed() {
    const now = nowText();
    await putRecord("users", defaultUser());
    await putRecord("accounts", defaultAccount());
    await this.seedDemoRecordsIfEmpty(now);
  }

  /** 首次使用且没有任何业务数据时，种入一套当月演示数据 */
  private async seedDemoRecordsIfEmpty(now: string) {
    const [txns, assets, budgets, loans, goals] = await Promise.all([
      getAllRecords<Transaction>("transactions"),
      getAllRecords<Asset>("assets"),
      getAllRecords<Budget>("budgets"),
      getAllRecords<Loan>("loans"),
      getAllRecords<SavingsGoal>("savingsGoals"),
    ]);
    if (txns.length || assets.length || budgets.length || loans.length || goals.length) return;

    const month = now.slice(0, 7);
    await putManyRecords("transactions", [
      { id: "txn-demo-salary", userId: "local-user", accountId: "default", amount: 12800, type: "INCOME", category: "工资", platform: "银行卡", merchant: "公司发薪", date: `${month}-15 09:00:00`, description: "月中发薪", createdAt: now, updatedAt: now },
      { id: "txn-demo-food", userId: "local-user", accountId: "default", amount: 86, type: "EXPENSE", category: "餐饮", platform: "支付宝", merchant: "午餐", date: `${month}-16 12:15:00`, description: "工作餐", createdAt: now, updatedAt: now },
      { id: "txn-demo-shop", userId: "local-user", accountId: "default", amount: 268, type: "EXPENSE", category: "购物", platform: "微信", merchant: "日用品", date: `${month}-18 20:10:00`, description: "家庭采购", createdAt: now, updatedAt: now },
      { id: "txn-demo-traffic", userId: "local-user", accountId: "default", amount: 48, type: "EXPENSE", category: "交通", platform: "支付宝", merchant: "地铁公交", date: `${month}-19 08:30:00`, description: "通勤", createdAt: now, updatedAt: now },
      { id: "txn-demo-entertain", userId: "local-user", accountId: "default", amount: 156, type: "EXPENSE", category: "娱乐", platform: "微信", merchant: "电影票", date: `${month}-20 20:40:00`, description: "周末观影", createdAt: now, updatedAt: now },
    ]);
    await putManyRecords("assets", [
      { id: "asset-demo-bank", userId: "local-user", accountId: "default", name: "工资卡", type: "BANK_CARD", balance: 48216.4, currency: "CNY", createdAt: now, updatedAt: now },
      { id: "asset-demo-wechat", userId: "local-user", accountId: "default", name: "微信钱包", type: "WECHAT", balance: 1260.5, currency: "CNY", createdAt: now, updatedAt: now },
      { id: "asset-demo-alipay", userId: "local-user", accountId: "default", name: "支付宝余额", type: "ALIPAY", balance: 3180, currency: "CNY", createdAt: now, updatedAt: now },
    ]);
    await putManyRecords("budgets", [
      { id: "budget-demo-global", userId: "local-user", accountId: "default", amount: 6000, category: "ALL", period: "MONTHLY", alertPercent: 80, platform: null, scopeType: "GLOBAL", createdAt: now, updatedAt: now },
      { id: "budget-demo-food", userId: "local-user", accountId: "default", amount: 1500, category: "餐饮", period: "MONTHLY", alertPercent: 80, platform: null, scopeType: "CATEGORY", createdAt: now, updatedAt: now },
    ]);
    await putManyRecords("loans", [
      { id: "loan-demo-home", userId: "local-user", accountId: "default", platform: "房贷", totalAmount: 480000, remainingAmount: 352000, periods: 240, paidPeriods: 64, monthlyPayment: 3200, dueDate: 20, status: "ACTIVE", matchKeywords: null, createdAt: now, updatedAt: now },
      { id: "loan-demo-car", userId: "local-user", accountId: "default", platform: "车贷", totalAmount: 80000, remainingAmount: 27000, periods: 36, paidPeriods: 18, monthlyPayment: 2200, dueDate: 10, status: "ACTIVE", matchKeywords: null, createdAt: now, updatedAt: now },
    ]);
    await putManyRecords("savingsGoals", [
      { id: "goal-demo-travel", userId: "local-user", accountId: "default", name: "旅行基金", targetAmount: 30000, currentAmount: 9200, deadline: "2026-12-31", type: "LONG_TERM", status: "ACTIVE", depositType: "CASH", planConfig: null, createdAt: now, updatedAt: now },
      { id: "goal-demo-emergency", userId: "local-user", accountId: "default", name: "应急储备", targetAmount: 20000, currentAmount: 6800, deadline: "2027-06-30", type: "LONG_TERM", status: "ACTIVE", depositType: "CASH", planConfig: null, createdAt: now, updatedAt: now },
    ]);
    await putManyRecords("savingsPlans", [
      { id: "plan-demo-travel-1", goalId: "goal-demo-travel", amount: 2000, status: "PENDING", month, createdAt: now, updatedAt: now },
    ]);
  }

  async getCurrentUser() {
    await this.ensureSeeded();
    return (await getRecord<User>("users", "local-user")) ?? null;
  }

  async saveUser(user: User) {
    await this.ensureSeeded();
    await putRecord("users", user);
  }

  async getAccounts() {
    await this.ensureSeeded();
    return getAllRecords<Account>("accounts");
  }

  async getAccount(id: string) {
    await this.ensureSeeded();
    return (await getRecord<Account>("accounts", id)) ?? null;
  }

  async saveAccount(account: Account) {
    await this.ensureSeeded();
    await putRecord("accounts", account);
  }

  async deleteAccount(id: string) {
    await this.ensureSeeded();
    await deleteRecord("accounts", id);
  }

  async getTransactions(accountId: string, page = 1, pageSize = 50) {
    await this.ensureSeeded();
    const targetAccountId = accountId || getCurrentAccountId();
    const all = sortByDateDesc(
      (await getAllRecords<Transaction>("transactions")).filter((item) => item.accountId === targetAccountId),
    );
    return all.slice((page - 1) * pageSize, page * pageSize);
  }

  async getTransaction(id: string) {
    await this.ensureSeeded();
    return (await getRecord<Transaction>("transactions", id)) ?? null;
  }

  async saveTransaction(transaction: Transaction) {
    await this.ensureSeeded();
    await putRecord("transactions", transaction);
  }

  async deleteTransaction(id: string) {
    await this.ensureSeeded();
    await deleteRecord("transactions", id);
  }

  async importTransactions(transactions: Transaction[]): Promise<ImportResult> {
    await this.ensureSeeded();
    const existing = await getAllRecords<Transaction>("transactions");
    const orderIdSet = new Set(existing.map((item) => item.orderId).filter(Boolean));
    const next = transactions.filter((item) => !item.orderId || !orderIdSet.has(item.orderId));
    await putManyRecords("transactions", next);
    return {
      imported: next.length,
      skipped: transactions.length - next.length,
      errors: 0,
    };
  }

  async getAssets(accountId: string) {
    await this.ensureSeeded();
    const targetAccountId = accountId || getCurrentAccountId();
    return (await getAllRecords<Asset>("assets")).filter((item) => item.accountId === targetAccountId);
  }

  async saveAsset(asset: Asset) {
    await this.ensureSeeded();
    await putRecord("assets", asset);
  }

  async deleteAsset(id: string) {
    await this.ensureSeeded();
    await deleteRecord("assets", id);
  }

  async getBudgets(accountId: string) {
    await this.ensureSeeded();
    const targetAccountId = accountId || getCurrentAccountId();
    return (await getAllRecords<Budget>("budgets")).filter((item) => item.accountId === targetAccountId);
  }

  async saveBudget(budget: Budget) {
    await this.ensureSeeded();
    await putRecord("budgets", budget);
  }

  async deleteBudget(id: string) {
    await this.ensureSeeded();
    await deleteRecord("budgets", id);
  }

  async getLoans(accountId: string) {
    await this.ensureSeeded();
    const targetAccountId = accountId || getCurrentAccountId();
    return (await getAllRecords<Loan>("loans")).filter((item) => item.accountId === targetAccountId);
  }

  async saveLoan(loan: Loan) {
    await this.ensureSeeded();
    await putRecord("loans", loan);
  }

  async deleteLoan(id: string) {
    await this.ensureSeeded();
    await deleteRecord("loans", id);
  }

  async getSavingsGoals(accountId: string) {
    await this.ensureSeeded();
    const targetAccountId = accountId || getCurrentAccountId();
    return (await getAllRecords<SavingsGoal>("savingsGoals")).filter((item) => item.accountId === targetAccountId);
  }

  async saveSavingsGoal(goal: SavingsGoal) {
    await this.ensureSeeded();
    await putRecord("savingsGoals", goal);
  }

  async deleteSavingsGoal(id: string) {
    await this.ensureSeeded();
    await deleteRecord("savingsGoals", id);
  }

  async getSavingsPlans(goalId: string) {
    await this.ensureSeeded();
    return (await getAllRecords<SavingsPlan>("savingsPlans")).filter((item) => item.goalId === goalId);
  }

  async saveSavingsPlan(plan: SavingsPlan) {
    await this.ensureSeeded();
    await putRecord("savingsPlans", plan);
  }

  async getCategoryRules(accountId: string) {
    await this.ensureSeeded();
    return (await getAllRecords<CategoryRule>("categoryRules")).filter((item) => item.accountId === accountId);
  }

  async saveCategoryRule(rule: CategoryRule) {
    await this.ensureSeeded();
    await putRecord("categoryRules", rule);
  }

  async getImportErrorLogs(accountId: string) {
    await this.ensureSeeded();
    return (await getAllRecords<ImportErrorLog>("importErrorLogs")).filter((item) => item.accountId === accountId);
  }

  async saveImportErrorLog(log: ImportErrorLog) {
    await this.ensureSeeded();
    await putRecord("importErrorLogs", log);
  }

  async getExchangeRates() {
    await this.ensureSeeded();
    return getAllRecords<ExchangeRate>("exchangeRates");
  }

  async getThemeConfig(userId: string) {
    await this.ensureSeeded();
    const configs = await getAllRecords<ThemeConfig>("themeConfigs");
    return configs.find((item) => item.userId === userId) ?? null;
  }

  async saveThemeConfig(config: ThemeConfig) {
    await this.ensureSeeded();
    await putRecord("themeConfigs", config);
  }
}
